### DISCUSSION TITLE
Encouraging and enforcing testing

### DISCUSSION THREAD

**Discussion by LeonRa** [MobileNativeFoundation/discussions#79] [General]
Hi everyone,

We are currently looking into encouraging more of our mobile developers to write tests at Duolingo. There is already a culture around this on the backend and mobile developers generally understand why doing so is important, but there are the expected trade-offs in priorities when it comes to shipping features. We'd love to know:

* How other companies encourage developers / feature teams to write tests
* What criteria other companies use for saying that something is sufficiently tested (e.g. coverage, user flows)
* What automation other companies use to enforce such criteria on a pull request (or similar) level

Thanks in advance!

**Comment by sberrevoets** [comment]
See also https://github.com/MobileNativeFoundation/discussions/discussions/9, some related discussions there including a few companies sharing answers to some of your questions.

**Reply by LeonRa** [reply]
Thanks @sberrevoets! There were definitely some great suggestions in there and in #6. Hoping for some more here as well. 🙏

**Comment by Koshub** [comment]
I think the first rule is to provide as much as possible information, training, etc to make sure the developers understand why tests are important and what are the benefits. You should make sure everybody understands that. It is not easy to do but doable.

Later teams can select different approaches to add more tests or start to add tests. It will depend on their skills to write tests. For example, add new code without any decrease in the code coverage report.

**Reply by LeonRa** [reply]
Thanks @Koshub, I definitely agree! We are at a point where the understanding is mostly there, and are now looking at those approaches for adding more tests. My hope is that the many wonderful folks here have run into this before have some strategies in place at their workplaces. I found https://github.com/MobileNativeFoundation/discussions/discussions/9#discussioncomment-423529 particularly helpful in this regard.

**Reply by Koshub** [reply]
Setting some rules on PR or pre-commit hooks works but I do not believe this way is really good enough. Let's say we are going to set the rule that the test coverage should be not less than 80%. What this number is based on? Why not 82%? I would say if developers "feel" the value of tests they keep trying to make the coverage near 100%. So the strategy of deciding what rules to set up should flow bottom up. The devs will find out what rules they need to chose to make sure they reach the value they would like to achieve. The coverage or other metrics are just tools to check and evaluate their results by themselves. So if you make some training course and after that the developers make their rules closer to 100% then the course was good. There is no need to push people to start to write tests or add more tests if they began to want to do it once they better understood the value.

**Reply by LeonRa** [reply]
That is a some very thoughtful advice for getting others to drive the discussion. Thank you for taking the time to write it out.

**Comment by nicolasrostan** [comment]
One thing I think helps us a lot is to make the templates for our modules with examples of testing, for example we have a template to generate a new pod for iOS, when a feature team wants to create a new module, it'll use it, and automatically get Quick and Nimble as a dependency, and a skeleton project with some tests.
Other things we do in CI are, set a coverage % for the module, and a % for the PR, in our case a module should have 80% of coverage but if it doesn't we just throw a warning, but if a PR doesn't have 80% on the new files it's adding then we'll block the PR. 
Of course educating through documentation and internal talks helps a lot too.

**Reply by LeonRa** [reply]
Thanks, this is great! We definitely plan to do more education here. How did you land on 80% as your baseline? Also, do you have any parts of your architecture you explicitly don't factor into this metric?

**Comment by omarzl** [comment]
Hello there! It is pretty interesting the point of view of Koshub, in our case our app grow too fast and we had a poor coverage because we were a startup focusing in generating features as crazy, now that we are getting mature as a company we were requested to increment coverage, since our current code doesn't have any test at all we decided to set our baseline to 60% for new files and 0% (for the moment) to existing ones since we can't block feature maintance

This first approach was to make the developers get used to write tests and then we will start increasing little by little the required coverage for existing files

Regarding specific tools we use slather and danger to gather the coverage and block the pr if necessary

**Reply by LeonRa** [reply]
Thank you for the response! We can definitely relate to the focus of growth and trying to stabilize things over time. We have a similar strategy planned now, though we're beginning with gentler automated nudges on PRs as opposed to blocking on specific coverage percentages. I'm curious about two things you mentioned:

1. How did you choose 60% as a baseline? Does that factor in things like Activities / Fragments or are some classes exempt?
2. How are you tracking which files are new vs existing over time? In other words, how do you make sure that a previously committed file with 60% coverage can't drop to below then when it is modified in the future?
